### PR TITLE
`LnkContestTestSuite` on backend

### PR DESCRIPTION
**Pull request by sanyavertolet**
This PR is connected with #1192

# **ALL EXISTING CONTESTS WILL LOSE THEIR TEST SUITES**
New ones won't

### What's done:
 * Added new table for links between contest and test suite
 * Added `LnkContestTestSuite` entity
 * Added Repository for `LnkContestTestSuite`
 * Added Service for `LnkContestTestSuite`
 * Removed `testSuiteIds` from `Contest`
 * Dropped column `test_suite_ids` from `Contest` table
 * Fixed existing endpoints to support new link entity

### CODE REVIEW

**Review by Cheshiriks — COMMENTED** [review]
You write that all contests will be lost. But you don't have a change set to clear your contest base.

**Comment by sanyavertolet** [comment]
Contest will remain, but information about test suites corresponding to contests will be lost
> You write that all contests will be lost. But you don't have a change set to clear your contest base.

**Comment by Cheshiriks** [comment]
> Contest will remain, but information about test suites corresponding to contests will be lost
> 
> > You write that all contests will be lost. But you don't have a change set to clear your contest base.

I'm sorry, I didn't read it right. Do I need to delete existing contests? Will they lead to errors?

**Comment by sanyavertolet** [comment]
> > Contest will remain, but information about test suites corresponding to contests will be lost
> > > You write that all contests will be lost. But you don't have a change set to clear your contest base.
> 
> I'm sorry, I didn't read it right. Do you need to delete existing contests? Will they lead to errors?

It should not cause any errors as far as the existing active contest has no test suites (nothing will be lost) and you cannot participate in finished contests

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Contest.kt [hunk 1]
```diff
@@ -31,46 +29,35 @@ class Contest(
     @ManyToOne
     @JoinColumn(name = "organization_id")
     var organization: Organization,
-    var testSuiteIds: String = "",
```
Comment (petertrr) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Contest.kt:34]: Can we have simple `@OneToMany var testSuites: List<TestSuite>` here?

Comment (sanyavertolet) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Contest.kt:34]: I am not sure how to map field in `Contest` entity and records in table `lnk_contest_test_suite`.
Seems like we need to store list of test suite ids in `contest` table, but as far as I remember mysql is unable to work with collections

Comment (Cheshiriks) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Contest.kt:34]: This does not require an test suite id in contest table. You can see how it is implemented @OneToMany in User entity

Comment (sanyavertolet) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/entities/Contest.kt:34]: Done

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 2]
```diff
@@ -408,17 +403,24 @@ internal class ContestController(
         }
         .map { (organization, contests) ->
             contestsRequest.map { contestRequest ->
-                contestService.updateContest(
-                    contestRequest.toContest(organization).apply { id = contests.single { name == it.get().name }.get().id }
-                )
+                contests.single {
+                    contestRequest.name == it.get().name && contestRequest.organizationName == it.get().organization.name
+                }
+                    .get()
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:409]: suggestion to change to `singleOrNull().orNotFound` to have an error in response

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 3]
```diff
@@ -408,17 +403,24 @@ internal class ContestController(
...
             }
             ResponseEntity.ok("Contest successfully updated")
         }
 
-    private fun getContestOrNotFound(contestName: String): Mono<Contest> = Mono.just(contestName)
-        .flatMap {
-            Mono.justOrEmpty(contestService.findByName(it))
-        }
+    private fun getContestOrNotFound(contestName: String): Mono<Contest> = Mono.justOrEmpty(
+        contestService.findByName(contestName)
```
Comment (nulls) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:422]: now it's not a lazy calculation. can we change it to `Mono.fromCallable` or `blockingToMono`

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/contests/ContestCreationComponent.kt [hunk 4]
```diff
@@ -234,7 +217,9 @@ private fun contestCreationComponent() = FC<ContestCreationComponentProps> { pro
                         inputTextFormRequired {
                             form = InputTypes.CONTEST_TEST_SUITE_IDS
                             conflictMessage = null
-                            textValue = testSuites.map { it.name }.sorted().joinToString(", ")
+                            textValue = contestDto.testSuites.map { it.name }
```
Comment (nulls) [inline save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/contests/ContestCreationComponent.kt:220]: maybe add version here? for example `${it.name} (${it.version})`

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 5]
```diff
@@ -327,7 +311,16 @@ internal class ContestController(
         .switchIfEmptyToResponseException(HttpStatus.CONFLICT) {
             "Contest with name [${contestDto.name}] already exists!"
         }
+        .zipWith(
+            testSuitesService.findTestSuitesByIds(contestDto.testSuites.map { it.requiredId() }).toMono()
```
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:315]: Ok, will do it

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:409]: But what if we have a batch of requests?
Let's say the first one crashes. What will happen with others?

Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:409]: Some of requests could be applied while `i`-th and the later won't

File: save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/contests/ContestCreationComponent.kt (hunk 4 above)
Comment (sanyavertolet) [reply save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/basic/contests/ContestCreationComponent.kt:220]: To my mind, the line will be TOO long to be readable

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 2 above)
Comment (nulls) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:409]: `single {}` returns Optional, `.get()` throws exception `NoSuchElementException`. I suggest to throw something meaningful

Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:409]: Makes sense, will do it

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 3 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt]: Changed

Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt]: Fixed

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt]: Done

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt [hunk 6]
```diff
@@ -190,7 +194,7 @@ internal class ContestController(
     ): Mono<TestFilesContent> = getContestOrNotFound(contestName)
         .zipWith(Mono.just(testSuiteId))
         .filter { (contest, testSuiteId) ->
-            testSuiteId in contest.getTestSuiteIds()
+            testSuiteId in lnkContestTestSuiteService.getAllTestSuiteIdsByContest(contest)
```
Comment (Cheshiriks) [inline save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt:197]: Is it necessary to have so many checks?
Why not just do:
`contest.testSuiteLinks.map { it.testSuite }.filter { testSuiteId in it.id }` ?

**Review by Cheshiriks — CHANGES_REQUESTED** [review]

File: save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt (hunk 6 above)
Comment (sanyavertolet) [reply save-backend/src/main/kotlin/com/saveourtool/save/backend/controllers/ContestController.kt]: A bit reworked this endpoint

**Review by Cheshiriks — APPROVED** [review]

**Pull request merged**